ArduCopter/control_safe_rtl.cpp

```diff
@@ -0,0 +1,147 @@
+#include "Copter.h"
+
+/*
+ * Init and run calls for Safe_RTL flight mode
+ *
+ * This code uses the SafeRTL path that is already in memory, and feeds it into WPNav, one point at a time.
+ * Once the copter is close to home, it will run a standard land controller.
+ */
+
+bool Copter::safe_rtl_init(bool ignore_checks)
+{
+    Vector3f current_pos;
+    if ((position_ok() || ignore_checks) && safe_rtl_path.is_active() && ahrs.get_relative_position_NED_origin(current_pos)) {
+        // initialise waypoint and spline controller
+        wp_nav->wp_and_spline_init();
+        wp_nav->init_loiter_target();
+
+        // if we're within 4m of the landing point, just skip right to pre-land-positioning
+        // otherwise, begin a normal safertl procedure from the start
+        if (HYPOT(current_pos, safe_rtl_path.get_point(0)) <= 4.0f) {
+            current_pos[2] -= 2.0f; // go to the point two meters above the landing spot, then start landing.
+            safe_rtl_state = SafeRTL_PreLandPosition;
+        } else {
+            safe_rtl_state = SafeRTL_WaitForPathCleanup;
+        }
+
+        // set current position as the target point.
+        DataFlash.Log_Write_SRTL(DataFlash_Class::SRTL_POINT_GOTO, current_pos);
+        wp_nav->set_wp_destination_NED_origin(current_pos);
+
+        // initialise yaw to obey user parameter
+        set_auto_yaw_mode(get_default_auto_yaw_mode(true));
+
+        // tell library to stop accepting new breadcrumbs
+        safe_rtl_path.accepting_new_points(false);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void Copter::safe_rtl_run()
+{
+    switch(safe_rtl_state){
```

Comment:
I modeled this after the way it was done in `control_rtl.cpp`. That used a switch statement and didn't have a default statement. I'm not sure what benefit the default statement would bring. I understand the "just in case" mentality, but I think if safe_rtl_state somehow ends up being a random value, we're probably screwed anyways.